### DISCUSSION TITLE
Alternative to GitHub Copilot

### DISCUSSION THREAD

**Discussion by rkrug** [posit-dev/positron#3871] [Q&A] [answered]
I understand that GitHub copilot can not be loaded into Positron because of licensing issues. But is there a good alternative to use, particularly for R?

**Comment by iainmwallace** [comment]
I like the continue extension - https://www.continue.dev

**Reply by rkrug** [reply]
Thanks - but this confuses me completely in setting it up.  will keep it in mind as it seems to offer many options.

**Comment by jthomasmock** [comment, marked as answer]
You can see most of the AI extensions by searching for "AI" at: Open VSX https://open-vsx.org/?search=AI&sortBy=downloadCount&sortOrder=desc OR in the Positron Extensions pane.

We would like to get support for GitHub Copilot in the future, but in the meantime there are many extensions that provide Chat, Inline completions, and other modalities of using GenAI.

Going purely off total downloads, not stack-ranked as I have not used all of them and am thus not vouching for any particular tool.

* GitLab Workflow (includes some AI features)
* Tabnine: Chat + Code completion
* Continue -- use many different local models or API endpoints
* Codeium: Chat + Code completion
* Cody by SourceGraph: Chat + Code completion
* Google Gemini: Chat + Code completion
* Amazon Q: Chat + Code completion
* Chat by GenieAI
* Huggingface LLMs for code completion

There are many other integrations for local models, OpenAI/ChatGPT, or various other as well. 

[Continue](https://open-vsx.org/extension/Continue/continue) for example has a nice integration and allows for running local models with no internet/API calls.

**Reply by rkrug** [reply]
Thanks - 
good to her that you are planning GitHub Copilot integration! 

I like the list, especially the info for each. I think I will try Tabbing for the moment.

**Comment by rkrug** [comment]
I have started using Tabbing, and it works nicely - I would say faster than GitHub Copilot. I only miss 4th Command-I to ask directly from the selected code.

**Reply by jthomasmock** [reply]
Great! I think you're using Tabnine. If so, you might check the shortcuts or use the Command Palette (Cmd/Ctrl + Shift + P) to focus code and ask a question about it:
![image](https://github.com/posit-dev/positron/assets/29187501/62a2969e-cab4-4d90-98ce-91016ec01098)

There's also a bit more in their docs: https://docs.tabnine.com/main/getting-started/quickstart#id-4.-getting-more-out-of-tabnine

**Reply by hafen** [reply]
Are you using tabnine for R or Python? I've been trying it for R and its suggestions are almost always useless and never what I have in mind, while GitHub Copilot seems to be able to read my mind.

**Reply by rkrug** [reply]
I have switched to Codeium - no particular reason.

**Comment by rkrug** [comment]
Thanks a lot - works perfectly.

**Comment by ntluong95** [comment]
I don't have subscription for TabNine. Any plan in near future that Copilot is going to be supported?

**Comment by rkrug** [comment]
I neither - but it works (at least it gives me answers which are coming close to co-pilot).

**Comment by SoaresAlisson** [comment]
I'm using codeium https://codeium.com/ It's free for individual users and has plugins to nvim, vscode as well as for Positron, and a lot of other IDEs, but not for Rstudio, so long as I know. It also works like magic with R, reading your thoughts and printing is as code.

**Reply by ntluong95** [reply]
Wow that's very good alternative. Free version even doesn't need to register

**Reply by SoaresAlisson** [reply]
I'm using it in nvim. It demands me to log in in the website and paste the API key in nvim console.

**Reply by taiyodayo** [reply]
This was super quick to setup! And has features like auto doc-string generation right in the editor.
I'm liking this more than copilot already.

**Comment by kv9898** [comment]
I would vote for Codeium

**Comment by davidzgli** [comment]
Have anybody tried Cline (previously Claude Dev)?  I just started.  It works insane to read your quarto script, and add a new section, such as compute summary statistics for a tibble grouped by a factor.

**Reply by davidzgli** [reply]
The problem I have with Cline is the "Shell Integration" is unavailable in Positron.  It makes Cline not be able to see the shell output, and cannot automate some jobs.   I am not sure if it is intentionally disabled in Position since "Shell Integration" seems to work in vscode. My vscode version is 1.93.0.
